ballista/scheduler/src/circuit_breaker/controller.rs

```diff
@@ -0,0 +1,167 @@
+use std::sync::Arc;
+
+use ballista_core::serde::protobuf::CircuitBreakerKey;
+use dashmap::DashMap;
+use tracing::{info, warn};
+
+#[derive(Clone)]
+pub struct CircuitBreakerController {
+    job_states: Arc<DashMap<String, JobState>>,
+}
+
+struct JobState {
+    stage_states: Arc<DashMap<u32, StageState>>,
+}
+
+struct StageState {
+    attempt_states: Arc<DashMap<u32, AttemptState>>,
+}
+
+struct AttemptState {
+    node_states: Arc<DashMap<String, NodeState>>,
+}
+
+struct NodeState {
+    partition_states: Arc<DashMap<u32, PartitionState>>,
+}
```

Comment:
I'm not 100% certain but my bet would be that locking at all of these levels will be worse than just wrapping a regular `HashMap` in a `parking_lot::RwLock`

Reply:
Yeah, maybe but my thinking when writing that was roughly that the entire operation should be fast so it probably isn't a huge bottleneck in general and even if it is, you are replacing acquiring one write lock with acquiring 4 read locks (and incrementing the ref count on each of the `Arc`s) and 1 write lock. The `DashMap` is sharded so you can in theory get less contention on each lock but it seems like it's probably not worth the tradeoff.

Reply:
Okay, done! :)